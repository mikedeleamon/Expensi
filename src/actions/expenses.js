import uuid from 'uuid'
import database from '../firebase/firebase'


//ACTIONS
//add
//const expense = {description,note,amount,createdAt}
export const addExpense = (expense) => (
    {type:'ADD_EXPENSE',
    expense:expense}
)
export const startAddExpense = (expenseData = {}) =>{ 
    return (dispatch,getState) => {
        const uid = getState().auth.uid
        const {description='',note='',amount=0,createdAt=0} = expenseData
        const expense = {description,note,amount,createdAt}

       return database.ref(`users/${uid}/expenses`)
       
        .push(expense)
        .then((ref)=>{
            dispatch(addExpense({
                id:ref.key,
                ...expense
            }))
        }).catch((error)=>{console.log(error)})
    }
}

//remove
export const removeExpense = ({id}={}) => (
    {type:'REMOVE_EXPENSE',
        id:id
    })

export const startRemoveExpense = ({id}={}) => {
    return(dispatch,getState) => {
        const uid = getState().auth.uid
    return database.ref(`users/${uid}/expenses/${id}`)
    .remove()
    .then(()=>{
        dispatch(removeExpense({id}))
    }).catch((error)=>{console.log(error)})
}
}

//edit
export const editExpense = (id ,updates) => ({
    type:'EDIT_EXPENSE',
    
        id:id,
        updates:updates
    
})

export const startEditExpense = (id, updates) => {
    return(dispatch,getState) => {
        const uid = getState().auth.uid
        return database.ref(`users/${uid}/expenses/${id}`)
        .update(updates)
        .then(()=>{
            dispatch(editExpense(id,updates))
        })
    }

}

export const setExpense = (expenses) => ({
    type:'SET_EXPENSE',
    expenses})

export const startSetExpenses = () => {
    return(dispatch,getState) =>{
        const uid = getState().auth.uid
    return database.ref(`users/${uid}/expenses`)
    .once('value')
    .then((snapshot)=> {
        const expenses = []
        
        snapshot.forEach((childSnapshot)=>{
            expenses.push({
                id:childSnapshot.key,
            ...childSnapshot.val()
        })
        })
        dispatch(setExpense(expenses))
    })
    }
}

