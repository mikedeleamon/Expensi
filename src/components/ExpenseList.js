import React from "react";
import {connect} from 'react-redux'
import ExpsenseListItem from "./ExpenseListItem";
import selectExpenses from '../selectors/expenses'


export const ExpenseList = (props) => (
    <div>
       <h1> Expense List</h1>
       {props.expenses.length === 0 ? (
        <p>no expense logged</p>
        ):(
        props.expenses.map((expense)=>{
        return( 
        <ExpsenseListItem key={expense.id} {...expense}/>
       )}
       )
       )}
    </div>
)

//set state for connect desired from the store
const mapStatetoProps = (state)=> {
    return {
        expenses:selectExpenses(state.expenses,state.filters)
    }
}

//connect takes in state
//needs provider already set up
export default connect(mapStatetoProps)(ExpenseList)

