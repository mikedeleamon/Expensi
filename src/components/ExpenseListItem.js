import React from 'react'
import moment from 'moment'
import numeral from 'numeral'
import {Link} from 'react-router-dom'


  const ExpsenseListItem = ({id, description, amount,createdAt}) => (
    
    <div>
      <div>
      <Link to={`/edit/${id}`}>
        <h2>{description}</h2>
      </Link>
      </div>
        <h3>{numeral(amount/100).format('$0,0.00')} - created at: {moment(createdAt).format('MMMM Do, YYYY')} </h3>
        
        
    </div>
  )

  export default ExpsenseListItem