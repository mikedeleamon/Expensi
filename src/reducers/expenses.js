const expenseStateDefault = []

export default  (state =expenseStateDefault, action) => {
    switch (action.type){
        case 'ADD_EXPENSE':
            //spread instead of using .concat()
            return [...state,action.expense]

        case 'REMOVE_EXPENSE':
            return state.filter(({id}) => id!==action.id)

        case 'EDIT_EXPENSE':
            return state.map((expense)=>{
                if (expense.id === action.id){
                    return{
                        ...expense,
                        //this will overide expenses
                        ...action.updates
                    }
                }else{
                    return expense
                }
            })
        default:
            return state
    }

}

