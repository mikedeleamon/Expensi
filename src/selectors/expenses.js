//Get visible expenses
import moment from 'moment'

export default (expenses, {text,sortBy,startDate,endDate}) =>(

    expenses.filter((expense)=> {
        //console.log(expense)
        const createdAtMoment = moment(expense.createdAt)
        const startDateMatch =  startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') :true
        const endDateMatch =  endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') :true
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase())
        return startDateMatch && endDateMatch && textMatch
    }).sort((objOne,objTwo)=>{
        if(sortBy==='date'){
            return objOne.createdAt < objTwo.createdAt ? 1 : -1
        } if(sortBy === 'amount'){
            return objOne.amount < objTwo.amount ? 1 : -1
        }
    })
)