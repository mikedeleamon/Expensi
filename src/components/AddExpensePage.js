import React from "react";
import expenses from "../selectors/expenses";
import {connect} from 'react-redux'
import ExpenseForm from "./ExpenseForm";
import { addExpense } from "../Actions/expenses";

export class AddExpensePage extends React.Component{
    addExpense=(expense)=>{
        this.props.addExpense(expense)
        {/*history allows auto navigation */}
        this.props.history.push('/')
    }
    render(){
        return(
        <div>
            <h1>Add an Expense</h1>
            <ExpenseForm onSubmit={this.addExpense}/>
        </div>
        )
    }
}


//another way to set up your dispatch calls
const mapDispatchToProps = (dispatch) => {
    return {
        addExpense:(expense) => dispatch(addExpense(expense))
    }
}

export default connect(undefined, mapDispatchToProps)(AddExpensePage)