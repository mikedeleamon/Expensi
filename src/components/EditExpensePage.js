import React from "react";
import {connect} from 'react-redux'
import { editExpense,removeExpense } from "../actions/expenses";
import ExpenseForm from "./ExpenseForm";

export class EditExpensePage extends React.Component{
    editExpense =(expense) => {
                this.props.editExpense(this.props.expense.id,expense)
                this.props.history.push('/')
            }
    
    removeExpense = () => {this.props.removeExpense({id:this.props.expense.id})
                            this.props.history.push('/')
                        }

    render(){
        return(
            <div> <h1>Edit an Expense</h1> {this.props.expense.id}
                <ExpenseForm 
                    expense= {this.props.expense}
                    onSubmit = { this.editExpense}

                />
                <button onClick ={this.removeExpense}>remove</button>
            </div>
        
        )
    }  
}

const matchDispacthToProps = (dispatch)=> {
    return{
        // allows you to remove the dispatch commanad by mapping it to props
        editExpense: (id, expense)=> dispatch(editExpense(id,expense)),
        removeExpense: (id) => dispatch(removeExpense(id))

    }
}
    

const mapStateToProps = (state,props) => {
    return{
        expense:state.expenses.find((expense)=> expense.id === props.match.params.id)   
        }
    }

export default connect (mapStateToProps,matchDispacthToProps)(EditExpensePage)