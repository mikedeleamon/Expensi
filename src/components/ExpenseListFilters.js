import React from "react";
import {connect} from 'react-redux'
import {DateRangePicker} from 'react-dates'
import { setFilterText,sortByAmount,sortByDate, setStartDate, setEndDate } from "../actions/filters";

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused:null
   }
    
    onDatesChange = ({startDate,endDate}) =>{
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    
    }

    onFocusChange = (calendarFocused) => (
        this.setState(()=>({calendarFocused}))
    )

    onSelectChange = (e)=> {e.target.value === 'date'? this.props.sortByDate():this.props.sortByAmount()}

    onTextChange = (e)  => {this.props.setFilterText(e.target.value)}

    render(){
    return(
    <div>
        <input type = 'text' placeholder="search" onChange ={this.onTextChange} value={this.props.filters.text}/>
        <select value={this.props.filters.sortBy} onChange = {this.onSelectChange}>
            <option value= 'date' >Date</option>
            <option value= 'amount' >Amount</option>
        </select>

    <DateRangePicker
        startDate={this.props.filters.startDate}
        endDate = {this.props.filters.endDate}
        onDatesChange = {this.onDatesChange}
        focusedInput= {this.state.calendarFocused}
        onFocusChange = {this.onFocusChange}
        numberOfMonths ={1}
        isOutsideRange = {()=>false}
        showClearDates = {true}
    />
    </div>
    )
    }

    


}
const mapStatetoProps = (state) =>(
    {
       filters:state.filters
    }
)

const mapDispatchToProps = (dispatch) => ({
    setFilterText: (text) => dispatch(setFilterText(text)),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    sortByAmount: () => dispatch(sortByAmount()),
    sortByDate: ()=> dispatch(sortByDate())
}
)

export default connect(mapStatetoProps,mapDispatchToProps)(ExpenseListFilters)