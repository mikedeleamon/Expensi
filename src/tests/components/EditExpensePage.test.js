import React from "react";
import {shallow} from 'enzyme'
import {EditExpensePage} from "../../components/EditExpensePage";
import expenses from '../fixtures/expenses'
import moment from 'moment'


let editExpenseSpy,removeExpenseSpy,history,expense,wrapper

beforeEach(()=>{
    editExpenseSpy = jest.fn()
    removeExpenseSpy = jest.fn()
    history = {push:jest.fn()}
    expense = expenses[1]
    wrapper = shallow(
    <EditExpensePage 
        startEditExpense = {editExpenseSpy} 
        startRemoveExpense={removeExpenseSpy} 
        history={history}
        expense = {expense}

        />
    )
})

test('should generate edit page properly',()=>{
    expect(wrapper).toMatchSnapshot()
})

test('should generate edit page properly with data',()=>{
    wrapper.find('ExpenseForm').prop('onSubmit')(expense)
    expect(editExpenseSpy).toHaveBeenLastCalledWith(expense.id,expense)
    expect(history.push).toHaveBeenLastCalledWith('/')
    expect(wrapper).toMatchSnapshot()
})

 test('should generate startRemoveExpense from expenses data',()=>{
    wrapper.find('button').simulate('click')
    expect(history.push).toHaveBeenLastCalledWith('/')
    expect(removeExpenseSpy).toHaveBeenLastCalledWith({id:expense.id})
     expect(wrapper).toMatchSnapshot()
 })