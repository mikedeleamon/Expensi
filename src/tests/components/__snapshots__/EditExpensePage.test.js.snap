// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should generate edit page properly 1`] = `
<div>
   
  <h1>
    Edit an Expense
  </h1>
   
  2
  <ExpenseForm
    expense={
      Object {
        "amount": 83300,
        "createdAt": -345600000,
        "description": "gas bill",
        "id": "2",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    remove
  </button>
</div>
`;

exports[`should generate edit page properly with data 1`] = `
<div>
   
  <h1>
    Edit an Expense
  </h1>
   
  2
  <ExpenseForm
    expense={
      Object {
        "amount": 83300,
        "createdAt": -345600000,
        "description": "gas bill",
        "id": "2",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    remove
  </button>
</div>
`;

exports[`should generate startRemoveExpense from expenses data 1`] = `
<div>
   
  <h1>
    Edit an Expense
  </h1>
   
  2
  <ExpenseForm
    expense={
      Object {
        "amount": 83300,
        "createdAt": -345600000,
        "description": "gas bill",
        "id": "2",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    remove
  </button>
</div>
`;
